### PR TITLE
Make parameter `pool` optional in CephFS storageclass

### PR DESCRIPTION
**Pull request by Daniel-Pivonka**
This makes the parameter `pool` optional in CephFS storageclass. When the 'pool' is provided the value is used and when it is not provided the default layout is used.

Fixes: #380 

Signed-off-by: Daniel-Pivonka <dpivonka@redhat.com>

### CODE REVIEW

**Comment by humblec** [comment]
@Madhu-1

File: examples/cephfs/storageclass.yaml [hunk 1]
```diff
@@ -17,7 +17,7 @@ parameters:
   # CephFS filesystem name into which the volume shall be created
   fsName: myfs
 
-  # Ceph pool into which the volume shall be created
+  # Ceph pool into which the volume shall be created (this is an optional parameter)
```
Comment (ShyamsundarR) [inline examples/cephfs/storageclass.yaml]: Reword: "(optional) Ceph pool into which volume data shall be stored (also change the string in deploy-cephfs.md)

Also comment the line below (line 21)

File: pkg/cephfs/volume.go [hunk 2]
```diff
@@ -74,6 +74,7 @@ func getVolumeRootPathCeph(volOptions *volumeOptions, cr *util.Credentials, volI
 }
 
 func createVolume(volOptions *volumeOptions, cr *util.Credentials, volID volumeID, bytesQuota int64) error {
+
```
Comment (ShyamsundarR) [inline pkg/cephfs/volume.go]: (nit) empty line

File: pkg/cephfs/volume.go [hunk 3]
```diff
@@ -104,12 +105,22 @@ func createVolume(volOptions *volumeOptions, cr *util.Credentials, volID volumeI
...
 		csiSubvolumeGroup,
-		"--pool_layout", volOptions.Pool,
-		"--mode", "777",
-		"-m", volOptions.Monitors,
-		"-c", util.CephConfigPath,
-		"-n", cephEntityClientPrefix+cr.ID,
-		"--keyfile="+cr.KeyFile)
+	}
+
+	if volOptions.Pool != "" {
```
Comment (ShyamsundarR) [inline pkg/cephfs/volume.go]: This option can be added to the end of the CLI args, so no need to split the command at exactly this point to append the same conditionally.

**Comment by ShyamsundarR** [comment]
@poornimag @ajarr PTAL as well, thanks.

**Comment by humblec** [comment]
@Daniel-Pivonka there are example yamls, may be update those too ?

File: e2e/utils.go [hunk 4]
```diff
@@ -203,7 +203,6 @@ func getSnapshot(path string) v1alpha1.VolumeSnapshot {
 func createCephfsStorageClass(c kubernetes.Interface, f *framework.Framework) {
 	scPath := fmt.Sprintf("%s/%s", cephfsExamplePath, "storageclass.yaml")
 	sc := getStorageClass(scPath)
-	sc.Parameters["pool"] = "myfs-data0"
```
Comment (Madhu-1) [inline e2e/utils.go]: as this is optional, please add a test case with and without pool parameter.

@poornimag is there a way we can identify in the backend (via toolbox pod)this image is having pool parameter enabled?

**Comment by Daniel-Pivonka** [comment]
> @Daniel-Pivonka there are example yamls, may be update those too ?

@humblec ive updated examples/cephfs/storageclass.yaml. Are there any others that need to be updated?

**Comment by humblec** [comment]
> > @Daniel-Pivonka there are example yamls, may be update those too ?
> 
> @humblec ive updated examples/cephfs/storageclass.yaml. Are there any others that need to be updated?

Thanks ! This need a follow up PR to rook as well, Isnt it ?  @Madhu-1

File: e2e/utils.go (hunk 4 above)
Comment (Daniel-Pivonka) [reply e2e/utils.go]: added a test for with and without the pool

**Comment by Daniel-Pivonka** [comment]
> > > @Daniel-Pivonka there are example yamls, may be update those too ?
> > 
> > 
> > @humblec ive updated examples/cephfs/storageclass.yaml. Are there any others that need to be updated?
> 
> Thanks ! This need a follow up PR to rook as well, Isnt it ? @Madhu-1

@humblec yes some changes in rook could be made to the example storageclass and the csi doc

File: e2e/utils.go [hunk 5]
```diff
@@ -200,11 +200,13 @@ func getSnapshot(path string) v1alpha1.VolumeSnapshot {
 	return sc
 }
 
-func createCephfsStorageClass(c kubernetes.Interface, f *framework.Framework) {
+func createCephfsStorageClass(c kubernetes.Interface, f *framework.Framework, enablePool bool) {
```
Comment (Madhu-1) [inline e2e/utils.go]: how about passing a map instead of a bool value. this function can be reused if someone wants to add new parameters
to storageclass

**Comment by ajarr** [comment]
> @poornimag @ajarr PTAL as well, thanks.

@ShyamsundarR , sorry missed this.

**Comment by ajarr** [comment]
LGTM

File: e2e/utils.go (hunk 5 above)
Comment (ShyamsundarR) [reply e2e/utils.go]: @Madhu-1 I would assume we can merge this in, without this change. When the next parameter needs to be added we can make this a map. Not scoring an approve on this as I do not want to merge without giving you an opportunity to get back on this.

**Comment by humblec** [comment]
LGTM. Approved.

**Pull request merged**